src/librustc_mir/transform/check_consts/qualifs.rs

```diff
@@ -0,0 +1,290 @@
+use rustc::mir::*;
+use rustc::mir::interpret::ConstValue;
+use rustc::ty::{self, Ty};
+use rustc_data_structures::bit_set::BitSet;
+use syntax_pos::DUMMY_SP;
+
+use super::Item as ConstCx;
+use super::validation::Mode;
+
+#[derive(Clone, Copy)]
+pub struct QualifSet(u8);
+
+impl QualifSet {
+    fn contains<Q: ?Sized + Qualif>(self) -> bool {
+        self.0 & (1 << Q::IDX) != 0
+    }
+}
+
+/// A "qualif"(-ication) is a way to look for something "bad" in the MIR that would disqualify some
+/// code for promotion or prevent it from evaluating at compile time. So `return true` means
+/// "I found something bad, no reason to go on searching". `false` is only returned if we
+/// definitely cannot find anything bad anywhere.
+///
+/// The default implementations proceed structurally.
+pub trait Qualif {
```

Comment:
Separating them completely lets me iterate a lot faster. I think they should be merged eventually, but maybe not in this PR.

I think there will be two significant changes from the original `Qualif`.
- The qualification state of each local gets its own parameter instead of being on `cx`.
- `cx` will store the `temp_promotion_state` (once we add `IsNotPromotable` and `IsNotImplicitlyPromotable` here).

There's other, smaller changes like using `validation::Mode` that doesn't have a `const fn` variant, but don't require big changes to the existing `qualify_consts.rs`.

Reply:
OK. Let's just remove the old version at some point instead of putting effort into deduplication